python-spec/src/somacore/query/query.py

```diff
@@ -434,6 +501,81 @@ def _axisp_inner(
         joinids = getattr(self._joinids, axis.value)
         return axisp[layer].read((joinids, joinids))
 
+    def _axism_inner(
+        self,
+        axis: "_Axis",
+        layer: str,
+    ) -> data.SparseRead:
+        key = axis.value + "m"
+
+        if key not in self._ms:
+            raise ValueError(f"Measurement does not contain {key} data")
+
+        axism = self._ms.obsm if axis is _Axis.OBS else self._ms.varm
+        if not (layer and layer in axism):
+            raise ValueError(f"Must specify '{key}' layer")
+
+        n_row, n_col = axism[layer].shape
+        col_joinids = pa.array(range(n_col))
+
+        joinids = getattr(self._joinids, axis.value)
+        return axism[layer].read((joinids, col_joinids))
+
+    def _axisp_inner_ndarray(
```

Comment:
all of these _inner_ndarray methods are largely identical to the _inner methods, with some processing done on the SparseRead result.

They should be DRY'ed.  For example, this one (I think) could be something like:

```
T = self._axisp_inner(axis, layer)
n_row = n_col = len(self._joinids.obs) if axis is _Axis.OBS else len(self._joinids.var)
idx = (self.indexer.by_obs if axis is _Axis.OBS else self.indexer.by_var)(T["soma_dim_0"])
Z = np.zeros(n_row * n_col, dtype=np.float32)
np.put(Z, idx * n_col + T["soma_dim_1"], T["soma_data"])
return Z.reshape(n_row, n_col)
```

Reply:
Should be way more DRY now.